### PR TITLE
[CN-291] Add hostPath for persistence

### PR DESCRIPTION
**Pull request by mtyazici**
Other changes:
- Add topologySpreadConstraints for spreading pods evenly across nodes
- Increase hostPath E2E test speed by reducing interval for reading logs from 1 sec to 10 ms.

### CODE REVIEW

File: controllers/hazelcast/validation/validation.go [hunk 1]
```diff
@@ -16,6 +16,10 @@ func ValidateSpec(h *hazelcastv1alpha1.Hazelcast) error {
 		return err
 	}
 
+	// if err := validatePersistence(h); err != nil {
+	// 	return err
+	// }
+
```
Comment (hakanmemisoglu) [inline controllers/hazelcast/validation/validation.go:22]: Commented section

File: controllers/hazelcast/validation/validation.go [hunk 2]
```diff
@@ -35,3 +39,10 @@ func validateLicense(h *hazelcastv1alpha1.Hazelcast) error {
 	}
 	return nil
 }
+
+// func validatePersistence(h *hazelcastv1alpha1.Hazelcast) error {
+// 	if h.Spec.Persistence.HostPath != "" &&  {
+// 		return errors.New("when Hazelcast Enterprise is deployed, licenseKeySecret must be set")
+// 	}
+// 	return nil
+// }
```
Comment (hakanmemisoglu) [inline controllers/hazelcast/validation/validation.go:48]: Same above

File: test/e2e/hazelcast_controller_e2e_test.go [hunk 3]
```diff
@@ -346,9 +353,13 @@ var _ = Describe("Hazelcast", func() {
...
 			By("Creating new Hazelcast cluster from existing backup")
 			baseDir := "/data/hot-restart/hot-backup/backup-" + seq
-			hazelcast = hazelcastconfig.PersistenceEnabled(hzNamespace, baseDir)
+			hazelcast = hazelcastconfig.PersistenceEnabled(hzNamespace, baseDir, useHostPath)
+			if useHostPath {
+				addNodeSelectorForName(hazelcast, nodeName)
+			}
 			Expect(k8sClient.Create(context.Background(), hazelcast)).Should(Succeed())
-			evaluateReadyMembers(lookupKey)
+			// TODO: Client sometimes cannot connect to servers, when its fixed we can use evaluateReadyMembers
```
Comment (hakanmemisoglu) [inline test/e2e/hazelcast_controller_e2e_test.go:361]: Can you revert it (after rebasing on the latest main commit) and try it one more time?

File: controllers/hazelcast/validation/validation.go (hunk 1 above)
Comment (mtyazici) [reply controllers/hazelcast/validation/validation.go:22]: I will remove them, thanks

File: controllers/hazelcast/validation/validation.go (hunk 2 above)
Comment (mtyazici) [reply controllers/hazelcast/validation/validation.go:48]: Here too

File: test/e2e/hazelcast_controller_e2e_test.go (hunk 3 above)
Comment (mtyazici) [reply test/e2e/hazelcast_controller_e2e_test.go:361]: I tried it with your fix and it worked. Maybe I should wait for you to merge client PR and rebase it.

**Review by hakanmemisoglu — APPROVED** [review]

File: controllers/hazelcast/hazelcast_resources.go [hunk 4]
```diff
@@ -532,7 +533,13 @@ func (r *HazelcastReconciler) reconcileStatefulset(ctx context.Context, h *hazel
 	}
 
 	if h.Spec.Persistence.IsEnabled() {
-		sts.Spec.VolumeClaimTemplates = persistentVolumeClaim(h)
+		if h.Spec.Persistence.UseHostPath() {
+			sts.Spec.Template.Spec.Volumes = append(sts.Spec.Template.Spec.Volumes, hostPathVolume(h))
+			sts.Spec.Template.Spec.Containers[0].SecurityContext.RunAsNonRoot = &[]bool{false}[0]
+			sts.Spec.Template.Spec.Containers[0].SecurityContext.RunAsUser = &[]int64{0}[0]
```
Comment (hasancelik) [inline controllers/hazelcast/hazelcast_resources.go:539]: Running pod with root privileges is not recommended due to security reasons. We need to inform users who prefer hostPath at our documentation. cc: @hakanmemisoglu 
We can share reference of the warning below:
https://kubernetes.io/docs/concepts/storage/volumes/#hostpath

**Review by hasancelik — APPROVED** [review]
LGTM 👍

**Pull request merged**